### ISSUE TITLE
Retrieve bound transaction in $beforeUpdate

### ISSUE THREAD

**Issue by jsneden** [Vincit/objection.js#59]
Hello,

I'm trying trigger a separate `insert` (based off the model) whenever an `update` occurs on a model. Obviously, I need this to be atomic, so I'm looking for a way to bind a transaction to the `update` and the `insert`.

Ideally, I'd create the transaction in the `$beforeUpdate` function and bind it to the current object as well as the additional insert object. However, that seems quite difficult to accomplish with the way `bindTransaction` works. So, alternatively, I'd like to bind a transaction to the `update` outside of the model class, and be able to retrieve that transaction object from within the `$beforeUpdate` function so I can create the `insert` statement with a bound transaction.

Of course, I realize that I could create the `insert` outside of the model (in the service in the context of the transaction) but I want to abstract this `insert` away from the service developer. Any information you can provide to help me accomplish an additional atomic action *within* the model is appreciated.

Thanks,
Jeff

**Comment by koskimas** [comment]
You can do this:

```js
SomeModel.prototype.$beforeUpdate = function () {
  var PossiblyBoundSomeModel = this.constructor;
  // The `knex` object and the `transaction` object are the same thing. Only difference
  // is that the transaction object holds on to the database connection between queries
  // instead of taking a new one from the pool. This works even if `PossiblyBoundSomeModel`
  // is not bound to a transaction.
  var tranasction = PossiblyBoundSomeModel.knex();
  return SomeOtherModel.bindTransaction(tranasction).query().insert({foo: 'bar'});
};
```

Does this solve your problem? This doesn't create the transaction in the `$beforeUpdate` but instead binds the new query to the (possibly) existing one.

**Comment by koskimas** [comment]
Maybe I should add a `$transaction()` method to model's prototype so that you we could write this:

SomeModel.prototype.$beforeUpdate = function () {
  return SomeOtherModel
    .bindTransaction(this.$transaction())
    .query()
    .insert({foo: 'bar'});
};

which is a lot cleaner. What do you think?

**Comment by jsneden** [comment]
I like that syntax, a lot. Does it require that a transaction is bound to `SomeModel` in the first place? It would be best if `this.$transaction` would bind itself to `this` if `this` does not currently have a transaction. That way the service which invoked `SomeModel.update()` could be oblivious to the transaction or pass one in if it is part of a greater transaction.

By the way, I still plan on testing your other suggestion, but I'm in the middle of refactoring my code to clean up how we write transactions. Thank you for the responses.

**Comment by koskimas** [comment]
Both of my suggestions would work even if there is no transaction bound to `SomeModel`, but they don't start a new transaction. In case there is no transaction bound to `SomeModel` the `bindTransaction` call would do nothing.

The problem with creating a transaction in the `$beforeUpdate` (or anywhere else for that matter) by calling the `this.$transaction()` is the ownership of the transaction Who is responsible for committing or rolling back the transaction? If a transaction is started by `this.$transaction()`, the original `update` operation should probably take ownership of the transaction. But what if a transaction already existed? You cannot add a `transaction.commit()` for example to `$afterUpdate` method because the `update` call may be a part of a larger transaction.

Also binding a model to a transaction in-place is not possible, since binding is done by creating a new Model subclass. What if you have 100 requests running in parallel that all create queries using a model class. Then you bind that model class to a transactions, you suddenly get 100s of queries joining your transaction and things go to hell. Model binding, the way done in objection, needs to be done by creating a new instance (subclass) of a model and binding that instance so that other parallel query chains using the same model are not affected.

I've been thinking about adding a `transacting` method to the query builder so that one could use transactions just like they are used in knex. I've been avoiding that because I have written some horrible bugs using that style of transaction handling, but the way it is currently done in objection is just as error-prone. Would you prefer the knex-style transaction over objection's way?

**Comment by jsneden** [comment]
Thanks, that explanation makes sense. Definitely a little confusing as to where the commit should occur (though perhaps `$afterUpdate` could check a status flag to see if this model initiated the transaction... and you still have to find a way to pass the transaction back to the service... it's getting pretty ugly at this point.)

I'm not too familiar with knex's transactions, but looking at the documentation, I don't see much difference between the way `transacting` works and the way `bindTransaction` works. Since you're curious, I'll add a few other of my thoughts on transactions at this point:

* I initially preferred the callback approach, but I was able to create a scenario that resulted in a non-atomic comi

**Comment by jsneden** [comment]
Thanks, that explanation makes sense. Definitely a little confusing as to where the commit should occur (though perhaps `$afterUpdate` could check a status flag to see if this model initiated the transaction... and you still have to find a way to pass the transaction back to the service... it's getting pretty ugly at this point.)

I'm not too familiar with knex's transactions, but looking at the documentation, I don't see much difference between the way `transacting` works and the way `bindTransaction` works. Since you're curious, I'll add a few other of my thoughts on transactions in objection at this point:

* I initially preferred the callback approach, but I was able to create a scenario that resulted in a non-atomic commit. The scenario I remember was an `insert` followed by an `update` where the `update` timed out because of a table lock. In this scenario or a server crash I was able to see the `insert` commit without the `update`.
* I switched to the transaction object approach, but I found it hard to externalize the inner functions and I hated that I had to keep writing back to local variables. We eventually settled on an approach where each function is externalized and depends on the transaction object being saved to the request object. The functions also share data through request.scope since the computed value is often needed outside of the immediately following function. See below.

```javascript
request.scope = {};
objection.transaction
	.start(Entity, Company)
	.then(function(transaction) { 
		DEBUG && console.log('DEBUG: Initiated transaction.');
		request.transaction = transaction;
	})

	.then(promiseCreateEntity.bind(this, request))
	.then(function(entity) {
		DEBUG && console.log('DEBUG: Created entity: ' + JSON.stringify(entity));
		request.scope.entity = entity;
	})

	.then(promiseReadUsedCodes.bind(this, request))
	.then(function(usedCodes) {
		DEBUG && console.log('DEBUG: Retrieved used codes: ' + JSON.stringify(usedCodes));
		request.scope.usedCodes = usedCodes;
	})

	.then(promiseCreateCompany.bind(this, request))
	.then(function(company) {
		DEBUG && console.log('DEBUG: Created company: ' + JSON.stringify(company));
		request.scope.company = company;
	})

	.then(promiseCreateUserCompany.bind(this, request))
	.then(function(userCompany) { 
		DEBUG && console.log('DEBUG: Related user to company: ' + JSON.stringify(userCompany));
		request.scope.userCompany = userCompany;
	})

	// Commit the transaction if everything was successful
	.then(function() {
		var scope = request.scope;
		var validator = revalidator.validate(scope, { properties: {
			company: { type: 'object', required: true }
		}});
		if ( ! validator.valid ) { throw new objection.ValidationError(validator); }

		response.send(201, scope.company);
		DEBUG && console.log('DEBUG: Committing transaction.');
		return request.transaction.commit();
	})

	// If any of the above queries throw an error, rollback the transaction
	.catch(function(error) {
		ERROR && console.log('ERROR: Company creation failed: ' + error);
		response.send(error.status || 500, error);
		if (request.transaction) {
			DEBUG && console.log('DEBUG: Rolling back transaction.');
			return request.transaction.rollback();
		}
		return;
	})
;

// Example promise function which expects the request to contain an authenticated user and a company object in request.scope
function promiseCreateUserCompany(request) {

	var validator = revalidator.validate(request, { properties: {
		user: {
			id: { type: 'integer' }
		},
		scope: {
			company: { type: 'object', required: true }
		}
	}});
	if ( ! validator.valid ) { throw new objection.ValidationError(validator); }

	return UserCompany
		.bindTransaction(request.transaction)
		.query()
		.insert({
			userEntityId: request.user.id,
			companyEntityId: scope.company.entityId,
			order: 0,
			role: 'OWNER'
		})
	;
}
```

**Comment by koskimas** [comment]
I usually use a very similar approach where I use the request as a scope. There really isn't an easier way to pass forward a large amount of variables using promise chains. This all gets a lot prettier with `async`/`await` or generators.

Have you used the bluebird `spread` method? For smaller chains it is quite handy. You can do stuff like this:

```js
Person
  .query()
  .where('firstName', 'Sylvester')
  .then(function (sylvester) {
    return [
      sylvester,
      sylvester.$relatedQuery('children').insert({firstName: 'Sage'})
    ];
  })
  .spread(function (sylvester, sage) {
    return [
      sylvester,
      sage, 
      sage.$relatedQuery('pets').insert({name: 'Fluffy'});
    ];
  })
  .spread(function (sylvester, sage, fluffy) {
    
  });
```

Now back to the topic. I will add the `$transaction` method for model instances and a `transaction` method for model classes. They can be used to fetch the current model's transaction like discussed above.

**Comment by koskimas** [comment]
Fixed in 470187cc8fc8a5c90ff5489046b3b3f87ab5272b
